File: benchmarks/wrappers/azure/python/storage.py

import os
import uuid

from azure.storage.blob import BlobServiceClient

class storage:
    instance = None
    client = None

    def __init__(self):
        self.client = BlobServiceClient.from_connection_string(
                os.getenv('STORAGE_CONNECTION_STRING')
            )

    @staticmethod
    def unique_name(name):
        name, extension = os.path.splitext(name)
        return '{name}.{random}{extension}'.format(
                    name=name,
                    extension=extension,
                    random=str(uuid.uuid4()).split('-')[0]
                )

    def upload(self, container, file, filepath, overwrite=False):
        with open(filepath, 'rb') as data:
            return self.upload_stream(container, file, data, overwrite)

    def download(self, container, file, filepath):
        with open(filepath, 'wb') as download_file:
            download_file.write( self.download_stream(container, file) )
    
    def download_directory(self, container, prefix, path):
        client = self.client.get_container_client(container=container)
        objects = client.list_blobs(name_starts_with=prefix)
        for obj in objects:
            file_name = obj.name
            path_to_file = os.path.dirname(file_name)
            os.makedirs(os.path.join(path, path_to_file), exist_ok=True)
            self.download(container, file_name, os.path.join(path, file_name))
    
    def upload_stream(self, container, file, data, overwrite=False):
        key_name = storage.unique_name(file)
        if (overwrite):
            key_name = file
        client = self.client.get_blob_client(
                container=container,
                blob=key_name
        )
        client.upload_blob(data, overwrite=overwrite)
        return key_name

    def download_stream(self, container, file):
        client = self.client.get_blob_client(container=container, blob=file)
        return client.download_blob().readall()
    
    def get_instance():
        if storage.instance is None:
            storage.instance = storage()
        return storage.instance

    def get_object(self, container, key):
        blob_client = self.client.get_blob_client(container=container, blob=key)
        downloader = blob_client.download_blob(max_concurrency=1, encoding='UTF-8')
        return downloader.readall()

    def list_blobs(self, container):
        client = self.client.get_container_client(container=container)

        # Azure returns an iterator. Turn it into a list.
        objs = []
        res = client.list_blob_names()
        for obj in res:
            objs.append(obj)

        return objs

File: benchmarks/wrappers/azure/nodejs/storage.js

const { BlobServiceClient } = require('@azure/storage-blob'),
        path = require('path'),
        uuid = require('uuid'),
        util = require('util'),
        stream = require('stream');


class azure_storage {

  constructor() {
    const STORAGE_CONNECTION_STRING = process.env.STORAGE_CONNECTION_STRING;   
    this.client = BlobServiceClient.fromConnectionString(STORAGE_CONNECTION_STRING);
  }

  unique_name(file) {
    let name = path.parse(file);
    let uuid_name = uuid.v4().split('-')[0];
    return path.join(name.dir, util.format('%s.%s%s', name.name, uuid_name, name.ext));
  }

  upload(container, file, filepath) {
    // it seems that JS does not have an API that would allow to
    // upload/download data without going through container/blob client 
    let containerClient = this.client.getContainerClient(container);
    let uniqueName = this.unique_name(file);
    let blockBlobClient = containerClient.getBlockBlobClient(uniqueName);
    return [uniqueName, blockBlobClient.uploadFile(filepath)];
  };

  download(bucket, file, filepath) {
    // TODO:
  };

  // We could provide additional API for just providing a byte buffer and uploading
  // Right now streams seems to be nicer
  uploadStream(container, file) {
    // it seems that JS does not have an API that would allow to
    // upload/download data without going through container/blob client 
    let containerClient = this.client.getContainerClient(container);
    let uniqueName = this.unique_name(file);
    let blockBlobClient = containerClient.getBlockBlobClient(uniqueName);
    var write_stream = new stream.PassThrough();
    // returns promise
    let upload = blockBlobClient.uploadStream(write_stream);
    return [write_stream, upload, uniqueName];
  };

  downloadStream(container, file) {
    let containerClient = this.client.getContainerClient(container);
    let blockBlobClient = containerClient.getBlockBlobClient(file);
    return blockBlobClient.download().then(
        (data) => {
          return data.readableStreamBody;
        }
    );
  };
};
exports.storage = azure_storage;
